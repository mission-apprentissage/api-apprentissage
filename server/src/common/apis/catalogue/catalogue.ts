import type { Readable } from "node:stream";

import { compose } from "oleoduc";
import { zFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

import logger from "@/common/logger";
import config from "@/config";

import { downloadFileInTmpFile } from "../../utils/apiUtils";
import { createJsonLineTransformStream } from "../../utils/streamUtils";
import getApiClient from "../client";

const catalogueClient = getApiClient(
  {
    baseURL: config.api.catalogue.baseurl,
  },
  { cache: false }
);

const neededFieldsFromCatalogue = Object.keys(zFormationCatalogue.shape).reduce<Record<string, number>>((acc, key) => {
  acc[key] = 1;
  return acc;
}, {});

export async function fetchCatalogueData(): Promise<Readable> {
  const query = JSON.stringify({ catalogue_published: true });

  const countResponse = await catalogueClient.get<number>("/api/v1/entity/formations/count", { params: { query } });

  logger.debug(`${countResponse.data} formation(s) Ã  importer du catalogue`);

  const response = await catalogueClient.get<Readable>("/api/v1/entity/formations.json", {
    responseType: "stream",
    params: {
      query,
      limit: countResponse.data,
      select: JSON.stringify(neededFieldsFromCatalogue),
    },
  });

  return compose(await downloadFileInTmpFile(response.data, "catalogue.zip"), createJsonLineTransformStream());
}
