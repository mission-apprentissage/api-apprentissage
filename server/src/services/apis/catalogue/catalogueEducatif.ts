import type { Readable } from "node:stream";

import axios from "axios";
import config from "@/config.js";
import { downloadFileAsStream } from "@/utils/apiUtils.js";
import { compose, createJsonLineTransformStream } from "@/utils/streamUtils.js";

const catalogueEducatifClient = axios.create({
  timeout: 5000,
  baseURL: config.api.catalogueEducatif.baseurl,
});

let cookieAuthCatalogueEducatif = "";

async function authCatalogueEducatif(): Promise<string | null> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const response = await catalogueEducatifClient.post<any>("/api/v1/auth/login", {
    username: config.api.catalogueEducatif.username,
    password: config.api.catalogueEducatif.password,
  });
  return response.headers["set-cookie"]?.join(";") || null;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function fetchFormationCatalogueEducatif(cle_ministere_educatif: string): Promise<any> {
  if (!cookieAuthCatalogueEducatif) {
    cookieAuthCatalogueEducatif = (await authCatalogueEducatif()) || "";
  }

  const query = JSON.stringify({
    cle_ministere_educatif,
    uai_formation_valide: true,
    $and: [{ uai_formation: { $ne: null } }, { uai_formation: { $ne: "" } }],
  });

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const response = await catalogueEducatifClient.get<any>("/api/v1/entity/formations", {
    headers: { cookie: cookieAuthCatalogueEducatif },
    params: {
      query,
      select: JSON.stringify({
        uai_formation: 1,
      }),
    },
  });

  if (!response.data.formations.length) return { uai_formation: null };

  return response.data.formations[0];
}

export async function fetchCatalogueEducatifData(): Promise<Readable> {
  if (!cookieAuthCatalogueEducatif) {
    cookieAuthCatalogueEducatif = (await authCatalogueEducatif()) || "";
  }
  const query = JSON.stringify({
    catalogue_published: true,
    uai_formation_valide: true,
    $and: [{ uai_formation: { $ne: null } }, { uai_formation: { $ne: "" } }],
  });

  const response = await catalogueEducatifClient.get<Readable>("/api/v1/entity/formations.json", {
    responseType: "stream",
    headers: { cookie: cookieAuthCatalogueEducatif },
    params: {
      query,
      select: JSON.stringify({
        cle_ministere_educatif: 1,
        uai_formation: 1,
      }),
      limit: 1000000,
    },
  });

  return compose(await downloadFileAsStream(response.data, "catalogueEducatif.zip"), createJsonLineTransformStream());
}
