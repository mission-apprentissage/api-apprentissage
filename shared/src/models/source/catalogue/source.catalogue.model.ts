import { z } from "zod";

import type { IModelDescriptorGeneric } from "../../common.js";
import { zObjectId } from "../../common.js";

const collectionName = "source.catalogue" as const;

const indexes: IModelDescriptorGeneric["indexes"] = [[{ date: 1, "data.cle_ministere_educatif": 1 }, {}]];

const etablissementFormateurSchema = z.object({
  etablissement_formateur_siret: z.string(),
  etablissement_formateur_uai: z.string().nullable(),
});

const etablissementGestionnaireSchema = z.object({
  etablissement_gestionnaire_siret: z.string(),
  etablissement_gestionnaire_uai: z.string().nullable(),
});

export const zFormationCatalogue = z
  .object({
    cle_ministere_educatif: z.string(),
    cfd: z.string(),
    code_postal: z.string(),
    code_commune_insee: z.string(),
    onisep_url: z.string().nullable(),
    onisep_intitule: z.string().nullable(),
    onisep_libelle_poursuite: z.string().nullable(),
    onisep_lien_site_onisepfr: z.string().nullable(),
    onisep_discipline: z.string().nullable(),
    onisep_domaine_sousdomaine: z.string().nullable(),
    rncp_code: z.string().nullable(),
    capacite: z.string().nullable(),
    duree: z.string(),
    annee: z.string(),
    email: z.string().nullable(),
    lieu_formation_geo_coordonnees: z.string(),
    lieu_formation_geo_coordonnees_computed: z.string().nullable(),
    lieu_formation_adresse: z.string(),
    lieu_formation_siret: z.string().nullish(),
    tags: z.array(z.string()),
    bcn_mefs_10: z.array(
      z.object({
        mef10: z.string(),
        modalite: z.object({ duree: z.string(), annee: z.string() }),
      })
    ),
    entierement_a_distance: z.boolean(),
    contenu: z.string().nullable(),
    objectif: z.string().nullable(),
    date_debut: z.array(z.string()).nullable(),
    date_fin: z.array(z.string()).nullable(),
    num_tel: z.string().nullable().describe("Numéro de téléphone de contact"),
    distance: z.number().nullable(),
    catalogue_published: z.boolean(),
    published: z.boolean(),
  })
  .merge(etablissementFormateurSchema)
  .merge(etablissementGestionnaireSchema);

export const zSourceCatalogue = z.object({
  _id: zObjectId,
  date: z.date(),
  data: zFormationCatalogue,
});

export type IFormationCatalogue = z.output<typeof zFormationCatalogue>;
export type ISourceCatalogue = z.output<typeof zSourceCatalogue>;

export default {
  zod: zSourceCatalogue,
  indexes,
  collectionName,
};
