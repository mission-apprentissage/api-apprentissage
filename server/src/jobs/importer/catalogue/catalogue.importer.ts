import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { internal } from "@hapi/boom";
import { ObjectId } from "mongodb";
import { zSourceCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

import { fetchCatalogueData } from "@/services/apis/catalogue/catalogue.js";
import { withCause } from "@/services/errors/withCause.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";
import { createBatchTransformStream } from "@/utils/streamUtils.js";

async function importCatalogueFormations(importDate: Date): Promise<void> {
  return pipeline(
    await fetchCatalogueData(),
    new Transform({
      objectMode: true,
      transform(data, _encoding, callback) {
        callback(
          null,
          zSourceCatalogue.parse({
            _id: new ObjectId(),
            date: importDate,
            data,
          })
        );
      },
    }),
    createBatchTransformStream({ size: 100 }),
    new Transform({
      objectMode: true,
      async transform(chunk, _encoding, callback) {
        try {
          await getDbCollection("source.catalogue").insertMany(chunk);
          callback();
        } catch (error) {
          callback(withCause(internal("import.catalogue: error when inserting"), error));
        }
      },
    })
  );
}

export async function runCatalogueImporter() {
  const importDate = new Date();
  const importId = new ObjectId();

  try {
    await getDbCollection("import.meta").insertOne({
      _id: importId,
      import_date: importDate,
      type: "catalogue",
      status: "pending",
    });

    await importCatalogueFormations(importDate);

    await getDbCollection("source.catalogue").deleteMany({
      date: { $ne: importDate },
    });

    await getDbCollection("import.meta").updateOne({ _id: importId }, { $set: { status: "done" } });

    return await getDbCollection("source.catalogue").countDocuments({ date: importDate });
  } catch (error) {
    await getDbCollection("import.meta").updateOne({ _id: importId }, { $set: { status: "failed" } });
    await getDbCollection("source.catalogue").deleteMany({ date: importDate });
    throw withCause(internal("import.catalogue: unable to runCatalogueImporter"), error);
  }
}
