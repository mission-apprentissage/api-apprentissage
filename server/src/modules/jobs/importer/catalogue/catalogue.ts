import { ObjectId } from "mongodb";
import { oleoduc, writeData } from "oleoduc";
import {
  IFormationCatalogue,
  IGeoPoint,
  zFormationCatalogue,
  zSourceCatalogue,
} from "shared/models/source/catalogue/source.catalogue.model";

import parentLogger from "@/common/logger";

import { getAllFormationsFromCatalogue } from "../../../../common/apis/catalogue/catalogue";
import { getDbCollection } from "../../../../common/utils/mongodbUtils";

const logger = parentLogger.child({ module: "import:catalogue" });

const convertStringCoordinatesToGeoPoint = (coordinates: string): IGeoPoint => {
  const coords = coordinates.split(",");

  return {
    type: "Point",
    coordinates: [parseFloat(coords[1]), parseFloat(coords[0])],
  };
};

const getCatalogueFormations = async () => {
  const stats = {
    total: 0,
    created: 0,
    failed: 0,
  };
  try {
    const formations = [] as IFormationCatalogue[];
    await oleoduc(
      await getAllFormationsFromCatalogue(),
      writeData(async (formation: any) => {
        stats.total++;
        try {
          // use MongoDB to add only add selected field from getAllFormationFromCatalogue() function and speedup the process
          delete formation._id; // break parsing / insertion otherwise
          formation.lieu_formation_geopoint = convertStringCoordinatesToGeoPoint(
            formation.lieu_formation_geo_coordonnees
          );
          const parsedFormation = zFormationCatalogue.parse(formation);
          formations.push(parsedFormation);
          stats.created++;
        } catch (e) {
          logger.error("Erreur enregistrement de formation", e);
          stats.failed++;
        }
      }),
      { parallel: 500 }
    );

    return { stats, formations };
  } catch (error) {
    // stop here if not able to get trainings (keep existing ones)
    logger.error(`Error fetching formations from Catalogue`, error);
    throw new Error("Error fetching formations from Catalogue");
  }
};

export async function runCatalogueImporter() {
  logger.info("Geting Catalogue ...");

  const importDate = new Date();

  const { stats, formations } = await getCatalogueFormations();

  const toInsert = formations.map((data) =>
    zSourceCatalogue.parse({
      _id: new ObjectId(),
      date: importDate,
      data,
    })
  );

  await getDbCollection("source.catalogue").insertMany(toInsert);

  await getDbCollection("source.catalogue").deleteMany({
    date: { $ne: importDate },
  });

  return {
    nb_formations: stats.created,
    erreurs: stats.failed,
  };
}
