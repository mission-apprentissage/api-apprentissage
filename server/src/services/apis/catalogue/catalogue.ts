import type { Readable } from "node:stream";

import { zFormationCatalogue } from "shared/models/source/catalogue/source.catalogue.model";

import axios from "axios";
import config from "@/config.js";
import logger from "@/services/logger.js";
import { downloadFileAsStream } from "@/utils/apiUtils.js";
import { compose, createJsonLineTransformStream } from "@/utils/streamUtils.js";

const catalogueClient = axios.create({
  baseURL: config.api.catalogue.baseurl,
  timeout: 60_000,
});

const neededFieldsFromCatalogue: Record<string, number> = Object.keys(zFormationCatalogue.shape).reduce<
  Record<string, number>
>((acc, key) => {
  acc[key] = 1;
  return acc;
}, {});

export async function fetchCatalogueData(): Promise<Readable> {
  const query = JSON.stringify({ catalogue_published: true });

  const countResponse = await catalogueClient.get<number>("/api/v1/entity/formations/count", { params: { query } });

  logger.debug(`${countResponse.data} formation(s) Ã  importer du catalogue`);

  const response = await catalogueClient.get<Readable>("/api/v1/entity/formations.json", {
    responseType: "stream",
    params: {
      query,
      limit: countResponse.data,
      select: JSON.stringify(neededFieldsFromCatalogue),
    },
  });

  return compose(await downloadFileAsStream(response.data, "catalogue.zip"), createJsonLineTransformStream());
}
