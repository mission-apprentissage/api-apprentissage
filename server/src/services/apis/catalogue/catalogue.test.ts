import nock, { cleanAll } from "nock";
import { Readable } from "stream";
import { afterEach, describe, expect, it } from "vitest";

import { fetchCatalogueData } from "./catalogue.js";

describe("getAllFormationsFromCatalogue", () => {
  afterEach(() => {
    cleanAll();
  });

  it("should return a Readable stream", async () => {
    const scopeCount = nock("https://catalogue-apprentissage.intercariforef.org")
      .get("/api/v1/entity/formations/count")
      .query({
        query: '{"published":true}',
      })
      .reply(200, "2");

    let scopeDataOptions = null;

    const scopeData = nock("https://catalogue-apprentissage.intercariforef.org")
      .get("/api/v1/entity/formations.json")
      .query((options) => {
        scopeDataOptions = options;
        return true;
      })
      .reply(200, JSON.stringify([{ id: "1" }, { id: 2 }]));

    const stream = await fetchCatalogueData();
    expect(stream).toBeInstanceOf(Readable);

    scopeCount.done();
    scopeData.done();

    expect(scopeDataOptions).toMatchSnapshot();

    const data = [];
    for await (const chunk of stream) {
      data.push(chunk);
    }

    expect(data).toEqual([{ id: "1" }, { id: 2 }]);
  });
});
