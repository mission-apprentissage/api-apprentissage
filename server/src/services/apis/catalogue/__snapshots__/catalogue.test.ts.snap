// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAllFormationsFromCatalogue > should return a Readable stream 1`] = `
{
  "limit": "2",
  "query": "{"catalogue_published":true}",
  "select": "{"cle_ministere_educatif":1,"cfd":1,"code_postal":1,"code_commune_insee":1,"onisep_url":1,"onisep_intitule":1,"onisep_libelle_poursuite":1,"onisep_lien_site_onisepfr":1,"onisep_discipline":1,"onisep_domaine_sousdomaine":1,"rncp_code":1,"capacite":1,"duree":1,"annee":1,"email":1,"lieu_formation_geo_coordonnees":1,"lieu_formation_geo_coordonnees_computed":1,"lieu_formation_adresse":1,"etablissement_lieu_formation_siret":1,"etablissement_lieu_formation_uai":1,"tags":1,"bcn_mefs_10":1,"entierement_a_distance":1,"contenu":1,"objectif":1,"date_debut":1,"date_fin":1,"num_tel":1,"distance":1,"catalogue_published":1,"published":1,"etablissement_formateur_siret":1,"etablissement_formateur_uai":1,"etablissement_gestionnaire_siret":1,"etablissement_gestionnaire_uai":1}",
}
`;
